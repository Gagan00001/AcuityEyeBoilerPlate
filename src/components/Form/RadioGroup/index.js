import React from "react";
import PropTypes from "prop-types";
import { Form, Radio } from "antd";
import cloneDeep from "lodash/cloneDeep";

import { idCalculator } from "../../../lib/util";

const RadioGroup = ({
  tabList,
  onChange,
  value,
  id,
  defaultValue,
  isFormItem,
  name,
  required,
  rules,
  ...otherProps
}) => {
  const clonedRules = cloneDeep(rules);
  if (required) {
    clonedRules.push({
      required: true,
      message: `${name} is required`,
    });
  }
  return (
    <div>
      {isFormItem ? (
        <Form.Item name={name} rules={[...clonedRules]} {...otherProps}>
          <Radio.Group
            value={value}
            buttonStyle="solid"
            onChange={onChange}
            defaultValue={defaultValue}
          >
            {tabList &&
              tabList.length > 0 &&
              tabList.map((item) => (
                <Radio.Button
                  id={idCalculator(id, item.value)}
                  value={item.value}
                  key={item.value}
                  data-testId={idCalculator("radio_group", item.value)}
                >
                  {item.label}
                </Radio.Button>
              ))}
          </Radio.Group>
        </Form.Item>
      ) : (
        <Radio.Group
          value={value}
          buttonStyle="solid"
          onChange={onChange}
          defaultValue={defaultValue}
        >
          {tabList &&
            tabList.length > 0 &&
            tabList.map((item) => (
              <Radio.Button
                id={idCalculator(id, item.value)}
                value={item.value}
                key={item.value}
                data-testId={idCalculator("radio_group", item.value)}
              >
                {item.label}
              </Radio.Button>
            ))}
        </Radio.Group>
      )}
    </div>
  );
};

RadioGroup.defaultProps = {
  tabList: [],
  onChange: () => {
    /* This is intentional */
  },
  value: "",
  defaultValue: "",
  id: "",
  isFormItem: false,
  rules: [],
  name: "",
  required: false,
  disabled: false,
};

RadioGroup.propTypes = {
  tabList: PropTypes.instanceOf(Array),
  onChange: PropTypes.func,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  isFormItem: PropTypes.bool,
  rules: PropTypes.instanceOf(Array),
  name: PropTypes.string,
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  defaultValue: PropTypes.string,
};

export default React.memo(RadioGroup);
