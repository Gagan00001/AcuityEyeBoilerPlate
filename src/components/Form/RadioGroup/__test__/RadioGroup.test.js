import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { unmountComponentAtNode } from 'react-dom';
import idCalculator from '../../../../lib/idCalculator';
import RadioGroup from '..';

let container = null;
const mockTabList = [
  { value: 'general', label: 'General' },
  { value: 'exception', label: 'Exceptions' },
  { value: 'modifiers', label: 'Modifiers' },
];

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('RadioGroup', () => {
  const props = {
    tabList: mockTabList,
    onChange: jest.fn(),
    value: 'general',
    labels: 'contracts/edit/terms',
    id: 'contract_terms_tabs',
  };

  const RenderRadioGroupComponent = () => (
    <RadioGroup {...props} />
  );

  test('should match snapshot', () => {
    const { container: Container } = render(
      <RenderRadioGroupComponent />, container,
    );
    expect(Container).toMatchSnapshot();
  });

  test('should render component', async () => {
    render(<RenderRadioGroupComponent />, container);
    mockTabList.forEach((tabList) => expect(screen.getByText(tabList.label)).toBeInTheDocument());
  });

  test('validate checked property of radio group', () => {
    render(<RenderRadioGroupComponent />, container);
    const radioGroupList = [];
    mockTabList.forEach((item, index) => {
      radioGroupList.push(screen.getByTestId(idCalculator('radio_group', item.value)));
      fireEvent.change(radioGroupList[index], { target: { checked: index === 1 } });
    });
    expect(radioGroupList[1].checked).toBe(true);
  });
});
