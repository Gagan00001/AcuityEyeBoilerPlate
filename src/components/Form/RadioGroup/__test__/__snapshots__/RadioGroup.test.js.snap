// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup should match snapshot 1`] = `
<div>
  <div>
    <div
      class="ant-radio-group ant-radio-group-solid"
    >
      <label
        class="ant-radio-button-wrapper ant-radio-button-wrapper-checked"
      >
        <span
          class="ant-radio-button ant-radio-button-checked"
        >
          <input
            checked=""
            class="ant-radio-button-input"
            data-testid="radio_group_general"
            id="contract_terms_tabs_general"
            type="radio"
            value="general"
          />
          <span
            class="ant-radio-button-inner"
          />
        </span>
        <span>
          General
        </span>
      </label>
      <label
        class="ant-radio-button-wrapper"
      >
        <span
          class="ant-radio-button"
        >
          <input
            class="ant-radio-button-input"
            data-testid="radio_group_exception"
            id="contract_terms_tabs_exception"
            type="radio"
            value="exception"
          />
          <span
            class="ant-radio-button-inner"
          />
        </span>
        <span>
          Exceptions
        </span>
      </label>
      <label
        class="ant-radio-button-wrapper"
      >
        <span
          class="ant-radio-button"
        >
          <input
            class="ant-radio-button-input"
            data-testid="radio_group_modifiers"
            id="contract_terms_tabs_modifiers"
            type="radio"
            value="modifiers"
          />
          <span
            class="ant-radio-button-inner"
          />
        </span>
        <span>
          Modifiers
        </span>
      </label>
    </div>
  </div>
</div>
`;
